clustering.py

```diff
@@ -244,3 +246,220 @@ def sort_labels(label_dict, cluster_labels, ts_to_labels):
+    num_clusters = (len(data) // 20) + 6
+    clusters_distances = []
+
+    for run_num in range(5):
+        centroids = k_means_init(data, num_clusters, run_num)
+
+        while True:
+            old_centroids = centroids.copy()
+
+            clusters, ts_to_cluster = update_clusters(data, centroids,
+                                                      must_link, can_not_link)
+            valid_clusters = update_centroids(data, clusters, centroids)
+
+            if not valid_clusters:
+                break
+
+            if np.array_equal(old_centroids, centroids):
+                assignment = [v for k, v in sorted(ts_to_cluster.items(),
+                                                   key=lambda item: item[0])]
+                center_dist = 0
+                for index, cluster in enumerate(assignment):
+                    center_dist += distance.euclidean(data[index],
+                                                      centroids[cluster])
+                clusters_distances.append([center_dist, assignment, run_num])
+                break
+
+    clusters_distances.sort()
+    return np.array(clusters_distances[0][1])
+
+def update_clusters(data, centroids, must_link, can_not_link):
+    """Updates the cluster assignments based on the centroids.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        centroids: A list of the centroids.
+        must_link: A dictionary mapping time series that must link.
+        can_not_link: A dictionary mapping time series that can't link.
+
+    Returns:
+        clusters: A list where the ith element is a list of the indexes
+            of the elements in the ith cluster.
+        ts_to_cluster: A dictionary mapping each time series to its
+            cluster assignment.
+    """
+    distances = pairwise_distances(data, centroids)
+    clusters = [[] for i in range(len(centroids))]
+    ts_to_cluster = {}
+
+    for ts_index, dist in enumerate(distances):
+        options = np.argsort(dist)
+        for option in options:
+            invalid = violates_cons(option, ts_index, ts_to_cluster,
+                                    must_link, can_not_link)
+            if not invalid:
+                ts_to_cluster[ts_index] = option
+                clusters[option].append(ts_index)
+                break
+    return clusters, ts_to_cluster
+
+def update_centroids(data, clusters, centroids):
+    """Updates the centroids based on the elements in each cluster.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        clusters: A list where the ith element is a list of the indexes
+            of the elements in the ith cluster.
+        centroids: A list of the centroids.
+
+    Returns:
+        True if the centroids were updated without an error, False
+        otherwise.
+    """
+    valid_clusters = True
```

Comment:
I don't think valid_clusters needs to be tracked here as a separate variable.

If the clusters are invalid, return False. Return also breaks out of the for loop.

As long as we're in the for loop we know the clusters are still valid.

If we iterate through everything without hitting the invalid conditional, simply return True.